doc/gaim-remote.1
.TH gaim-remote 1

.SH "NAME"
gaim-remote v0.66 \- Gaim command-line remote control

.SH "SYNOPSIS"
.TP 5
\fBgaim-remote \fI[command]\fR \fI[parameters]\fR
.TP 5
\fBgaim-remote -h \fI[command]\fR

.SH "DESCRIPTION"
.PP
\fbgaim-remote\fR is a command-line interface for controlling an already-running instance of the Gaim Instant Messaging client. It currently has only rudimentary support for launching AIM-style URIs, and closing the running instance, although further functionality is planned.

.SH "OPTIONS"
The following options are provided by \fBgaim-remote\fR:
.TP
.B \-h [\fIcommand\fB]
Display help for \fBgaim-remote\fR or a specific command (see below).

.SH "COMMANDS"
The following commands are implemented by \fBgaim-remote\fR:
.TP
.B quit
Closes the currently-running instance of \fBGaim\fR.
.TP
.B uri [URI]
Sends the given AIM-style URI to the currently running instance of \fBGaim\fR (see below).

.SH "URIs"
The following URIs are currently valid. These default to acting on the first AIM account you have in your list:
.TP
.B aim:goim?screenname=[SCREENNAME]
Opens a new conversation window to the given screenname.
.TP
.B aim:goim?screenname=[SCREENNAME]&message=[MESSAGE]
Sends the given message to the screenname. Spaces in the message must be replaced with +.
.TP
.B aim:gochat?roomname=[CHATROOM]
Opens a chat window in the given room.
.TP
.B aim:addbuddy?screenname=[SCREENNAME]
Opens a dialog to adds the specified buddy to your buddy list.

.SH "FILES"
.TP
.I /tmp/gaim_[USERNAME].[INSTANCE]
The UNIX domain socket used to contact the running instance of \fBGaim\fR.

.SH "SEE ALSO"
.BR gaim (1),
.BR \fIhttp://gaim.sourceforge.net/\fR,
.BR \fIhttp://www.sourceforge.net/projects/gaim/\fR

.SH "AUTHOR"
.PP
This manpage was written by Robert McQueen <\fIrobot101@debian.org\fR>
for the Debian GNU/Linux system (but may be used by others).



NEWS
-=[ Gaim ]=-  The Pimpin' Penguin IM Client That's Good For The Soul!
0.72 (10/30/2003):
	Sean: I think I'll go as the Gaim logo guy for Halloween.  That
	would be a cool costume.

	Luke: We've had a record month for number of bugs submitted (395 as
	of now) this month, but the total number of bugs open has grown by
	less than 20, so while it might be fustrating for those of you out
	there who think your bug report is being ignored, know that we are
	doing our best to look at them all :-). This release fixes a goodly
	number of bugs including the troubles with signing on icq (thanks
	to Mark). Enjoy!

	Christian: Last release brought forward a lot of bug reports, such
	as the ICQ bug, and also trouble with installing SSL. I can't count
	how many people were confused as to why their MSN plugin wasn't
	loading. Now, it'll load regardless, and when you try to connect,
	it'll give you an error dialog saying to install SSL (if you don't
	have it installed, of course). I didn't do a whole lot this release,
	but that'll change with the next. There are some fun things coming.

	Ethan: In complete defiance of history, I did something this
	release.  It had to do with IRC.  If I type here for long
	enough, I'm sure I'll remember what it was...  Oh yeah, all of
	you paranoid people can hide your local username by changing
	your IRC account settings.  I also want to note that we are
	aware that building Gaim is Harder than it Should Be on
	Solaris, and I am seeking solutions to some of the issues
	there.  Everyone remember to give some love to the developers
	who actually did something of note this release (i.e. not me)!

0.71 (10/09/2003):
	Rob: It's been a while since I've commited some news.  I've had a
	very hectic few weeks/months.  Thing are finally coming together
	again for me.  This release features much copying and pasting. Yeah.

	Sean: I can't wait for everything to be core/ui split!

	Christian: This is a good release, and I don't think people will have
	too many problems. There are a lot of little nice feature additions,
	and much progress has been made on the core/UI split. We should be
	done in a couple of releases, assuming we can get the remaining two or
	three parts core/UI-split. Oh, and plugin authors are going to hate me
	again, as a lot of code (specifically, the conversations code and utility
	functions) changed names. This is a Good Thing (TM), as always, but
	may require modifications to some plugins. Not all, fortunately, but
	some :) Many SSL problems were fixed, though we still have some to go.
	If you don't have SSL, you now need it for MSN. MSN won't even load
	without it. Starting October 15th, the old protocol won't be supported
	anymore, so there's no reason to support the old one anymore.

	Herman: I bet some plugin authors have resorted to making voodoo dolls
	of Christian.  Nothing new for Win Gaim users other than a fix for
	missing aim buddy icons, and removal of gtk-wimp (until it matures).

	Mark: People are going to make a voodoo doll of me and shower it with
	kisses and chocolate women: When an AIM user is away, you can now view
	the person's away message in their tooltip.  This is very similar to
	what iChat does, only our tooltip will wrap lines rather than displaying
	off your screen :-)

	Nathan: Chocolate women are cool, but I'm a bigger fan of the real kind.
	I re-wrote the Jabber support from scratch, so now there's at least one
	person who knows what's going on with it.  I got lost in the old code,
	even in the parts I wrote.  It now supports the SASL login stuff from
	the XMPP protocol specs, and will automagically upgrade the conncetion
	to SSL if it's supported.  If anything doesn't work like it did before,
	it's probably better, and you should be thankful for it.

	Ethan: I'm sure I'll be showered with all kinds of wonderful things,
	but I doubt it's because I did anything.  The reason being that I
	really don't think I did anything.  I meant to do some gaim_
	prepension and documentation for IRC and Tcl, and handle IRC errors
	more responsibly, but I've been meaning to do that for months.
	Except the Tcl part, the Tcl plugin hasn't even existed for months.
	I figure if I write enough NEWS no one will notice I didn't do
	anything during the release cycle.  Oh, I did close some bugs.  I
	didn't fix them, mind you, I just closed them.  With abandon.
	Don't tell Luke.

	Luke: Mmmm, I probly did even less than Ethan, except I closed rather
	more bugs for other reasons. Lots of you submitting duplicate bugs.
	This should be a pretty good release. We still have a couple crashes
	in ssl stuff, but we fixed a number of the bugs in the tracker, and
	the code cleaning helps alot as well. Enjoy!

0.70 (9/28/2003):
	Sean: Yahoo! works (for now).  That's good NEWS if I've ever heard it.  
	Huge thanks to our dear friends over at Cerulean Studios, creators of
	Trillian, for helping us out.

	Nathan: This release also works out most of the kinks in the new
	contact support.  The Jabber goodies I promised will have to wait
	for now.

	Christian: I didn't really do anything except fix a couple of issues
	loading protocol plugins that had plugin dependencies and getting perl
	to install where we tell it to. I had hoped that 0.70 would include
	a finished libgaim, but Yahoo kind of wrecked that dream. We'll see it
	within the next few releases. Life has been busy in a good way :)

0.69 (9/24/2003):
	Sean: MSN and Yahoo! work.  That's good NEWS if I've ever heard it.

	Herman: Win Gaim has been around for more than a year now.. Time Flies.

	Nathan: {{Meta,}{Contact,Buddy},Person} support for all!  Also, SSL
	support for Jabber makes its debut.  More Jabber goodies next release.

	Christian: It's a good release. The main highlight really is
	Nathan's Contact support, which is just beautiful. My buddy list
	shrunk considerably. As Sean stated, MSN and Yahoo! work.  MSN
	does not support the new MSN buddy icons and stuff, unfortunately,

src/gaim-remote.py
#!/usr/bin/python

import dbus
import re
import urllib
import sys

import xml.dom.minidom 

xml.dom.minidom.Element.all   = xml.dom.minidom.Element.getElementsByTagName

obj = dbus.SessionBus().get_object("net.sf.gaim.GaimService", "/net/sf/gaim/GaimObject")
gaim = dbus.Interface(obj, "net.sf.gaim.GaimInterface")

class CheckedObject:
    def __init__(self, obj):
        self.obj = obj

    def __getattr__(self, attr):
        return CheckedAttribute(self, attr)

class CheckedAttribute:
    def __init__(self, cobj, attr):
        self.cobj = cobj
        self.attr = attr
        
    def __call__(self, *args):
        result = self.cobj.obj.__getattr__(self.attr)(*args)
        if result == 0:
            raise "Error: " + self.attr + " " + str(args) + " returned " + str(result)
        return result
            
cgaim = CheckedObject(gaim)

urlregexp = r"^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?"

def extendlist(list, length, fill):
    if len(list) < length:
        return list + [fill] * (length - len(list))
    else:
        return list

def convert(value):
    try:
        return int(value)
    except:
        return value

def findaccount(accountname, protocolname):
    try:
        # prefer connected accounts
        account = cgaim.GaimAccountsFindConnected(accountname, protocolname)
        return account
    except:
        # try to get any account and connect it
        account = cgaim.GaimAccountsFindAny(accountname, protocolname)
        print gaim.GaimAccountGetUsername(account)
        gaim.GaimAccountSetStatusVargs(account, "online", 1)
        gaim.GaimAccountConnect(account)
        return account
    

def execute(uri):
    match = re.match(urlregexp, uri)
    protocol = match.group(2)
    if protocol == "aim" or protocol == "icq":
        protocol = "oscar"
    if protocol is not None:
        protocol = "prpl-" + protocol
    command = match.group(5)
    paramstring = match.group(7) 
    params = {}
    if paramstring is not None:
        for param in paramstring.split("&"):
            key, value = extendlist(param.split("=",1), 2, "")
            params[key] = urllib.unquote(value)

    accountname = params.get("account", "")

    if command == "goim":
        print params
        account = findaccount(accountname, protocol)
        conversation = cgaim.GaimConversationNew(1, account, params["screenname"])
        if "message" in params:
            im = cgaim.GaimConversationGetImData(conversation)
            gaim.GaimConvImSend(im, params["message"])
        return None

    elif command == "gochat":
        account = findaccount(accountname, protocol)
        connection = cgaim.GaimAccountGetConnection(account)
        return gaim.ServJoinChat(connection, params)

    elif command == "addbuddy":
        account = findaccount(accountname, protocol)
        return cgaim.GaimBlistRequestAddBuddy(account, params["screenname"],
                                              params.get("group", ""), "")

    elif command == "setstatus":
        if "account" in params:
            accounts = [cgaim.GaimAccountsFindAny(accountname, protocol)]
        else:
            accounts = gaim.GaimAccountsGetAllActive()

        for account in accounts:
            status = cgaim.GaimAccountGetStatus(account, params["status"])
            for key, value in params.items():
                if key not in ["status", "account"]:
                    gaim.GaimStatusSetAttrString(status, key, value)
            gaim.GaimAccountSetStatusVargs(account, params["status"], 1)
        return None

    elif command == "getinfo":
        account = findaccount(accountname, protocol)
        connection = cgaim.GaimAccountGetConnection(account)
        return gaim.ServGetInfo(connection, params["screenname"])

    elif command == "quit":
        return gaim.GaimCoreQuit()

    elif command == "uri":
        return None

    else:
        match = re.match(r"(\w+)\s*\(([^)]*)\)", command)
        if match is not None:
            name = match.group(1)
            argstr = match.group(2)
            if argstr == "":
                args = []
            else:
                args = argstr.split(",")
            fargs = []
            for arg in args:
                fargs.append(convert(arg.strip()))
            return gaim.__getattr__(name)(*fargs)
        else:
            # introspect the object to get parameter names and types
            # this is slow because the entire introspection info must be downloaded
            data = dbus.Interface(obj, "org.freedesktop.DBus.Introspectable").\
                   Introspect()
            introspect = xml.dom.minidom.parseString(data).documentElement
            for method in introspect.all("method"):
                if command == method.getAttribute("name"):
                    methodparams = []
                    for arg in method.all("arg"):
                        if arg.getAttribute("direction") == "in":
                            value = params[arg.getAttribute("name")]
                            type = arg.getAttribute("type")
                            if type == "s":
                                methodparams.append(value)
                            elif type == "i":
                                methodparams.append(int(value))
                            else:
                                raise "Don't know how to handle type \"%s\"" % type
                    return gaim.__getattr__(command)(*methodparams)
            raise "Unknown command: %s" % command


if len(sys.argv) == 1:
    print """This program uses DBus to communicate with gaim.

Usage:

    %s "command1" "command2" ...

Each command is of one of the three types:

    [protocol:]commandname?param1=value1&param2=value2&...
    FunctionName?param1=value1&param2=value2&...
    FunctionName(value1,value2,...)

The second and third form are provided for completeness but their use
is not recommended; use gaim-send or gaim-send-async instead.  The
second form uses introspection to find out the parameter names and
their types, therefore it is rather slow.

Examples of commands:

    jabber:goim?screenname=testone@localhost&message=hi
    jabber:gochat?room=TestRoom&server=conference.localhost
    jabber:getinfo?screenname=testone@localhost
    jabber:addbuddy?screenname=my friend

    setstatus?status=away&message=don't disturb
    quit

    GaimAccountsFindConnected?name=&protocol=prpl-jabber
    GaimAccountFindConnected(,prpl-jabber)
""" % sys.argv[0]

for arg in sys.argv[1:]:
    print execute(arg)
    
    
